### PR TITLE
Implementation of CCPP-ized TJ2016

### PR DESCRIPTION
**Pull request by mwaxmonsky**
Implements TJ2016 schemes using CCPP-ized metadata files.  Required the following changes:
1) Updated Kessler/ZM/geopotential meta data files to use updated gravity standard name
2) Added test suite definition files to test the before/after coupler
functions in isolation
3) Updated the design of the `run` functions to pass in state data and output
tendencies (instead of modifying state directly) and the tendencies are now
applied afterwards via additional `apply_*` schemes listed in the suite
definition file.
4) Updates several variables from CAM that were 2D to now be 3D to enable
future updates to take the vertical dimension into account for different
planetary physics.
5) Adds `index_top_interface` in `*_sfc_pbl_hs` scheme to allow integrating
with models that invert the vertical dimension data.
6) Introduces minor answer change to `*_sfc_pbl_hs` functionality.

As mentioned, there is a single know change to the after coupler run function
(tj2016_sfc_pbl_hs) which has consistently been off on timestep 2 only
by at most `0.13E-22` for a single variable (excerpt from run/atm log):
```
********** Physics Check Data Results **********

TIMESTEP:            2

 Variable                                          # Diffs  Max Diff   Max Diff loc (rank, col, lev)
 --------                                          -------  --------   -----------------------------
water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water
                                                         3  0.13E-22   (43,7,29)

********** End Physics Check Data Results **********
```

### CODE REVIEW

**Review by nusbaume — CHANGES_REQUESTED** [review]
Thanks for all the hard work in getting this through @mwaxmonsky!  I do have some change requests, but hopefully all of them are fairly minor.  Of course if you have any questions or concerns with any of my requests just let me know!

File: kessler/kessler_update.meta [hunk 1]
```diff
@@ -5,7 +5,7 @@
   name = kessler_update_init
   type = scheme
 [ gravit_in ]
-  standard_name = gravitational_acceleration
+  standard_name = standard_ravitational_acceleration
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply kessler/kessler_update.meta]: Good catch!  Fixed!

File: tj2016/suite_tj2016_precip.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<suite name="tj2016_precip" version="1.0">
+  <group name="physics_before_coupler">
+    <scheme>tj2016_precip</scheme>
+    <scheme>apply_heating_rate</scheme>
+    <scheme>qneg</scheme>
+  </group>
+</suite>
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/suite_tj2016_precip.xml]: Moved!

File: tj2016/suite_tj2016_sfc_pbl_hs.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/suite_tj2016_sfc_pbl_hs.xml]: Moved!

File: tj2016/tj2016_precip.F90 [hunk 4]
```diff
@@ -0,0 +1,117 @@
+module TJ2016_precip
+
+    use ccpp_kinds, only: kind_phys
+    use shr_const_mod, only: pi => shr_const_pi
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.F90]: Removed and updated meta.

File: tj2016/tj2016_precip.F90 [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  
+    implicit none
+    private
+    save
+
+    public :: tj2016_precip_run
+
+CONTAINS
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.F90]: Lowercased!

File: tj2016/tj2016_precip.F90 [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        pmid, pdel, T, qv, relhum, precl, tendency_of_air_enthalpy, scheme_name, errmsg, errflg)
+        !------------------------------------------------
+        !   Input / output parameters
+        !------------------------------------------------
+
+        integer,  intent(in)    :: ncol                    ! number of columns
+        integer,  intent(in)    :: pver                    ! number of vertical levels
+        
+        real(kind_phys), intent(in)    :: gravit      ! g: gravitational acceleration (m/s2)
+        real(kind_phys), intent(in)    :: cappa       ! Rd/cp
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.F90]: Removed!  It's only used in `sfc_pbl_hs` so I must have copied it into both without looking closer.

File: tj2016/tj2016_precip.F90 [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        real(kind_phys), intent(in)    :: rairv(:,:)  ! Rd: dry air gas constant (J/K/kg)
+        real(kind_phys), intent(in)    :: cpairv(:,:) ! cp: specific heat of dry air (J/K/kg)
+        real(kind_phys), intent(in)    :: latvap      ! L: latent heat of vaporization (J/kg)
+        real(kind_phys), intent(in)    :: rh2o        ! Rv: water vapor gas constant (J/K/kg)
+        real(kind_phys), intent(in)    :: epsilo      ! Rd/Rv: ratio of h2o to dry air molecular weights
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.F90:34]: Removed!

File: tj2016/tj2016_precip.F90 [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        real(kind_phys), intent(in)    :: dtime                   ! time step (s)
+        real(kind_phys), intent(in)    :: pmid(:,:)         ! mid-point pressure (Pa)
+        real(kind_phys), intent(in)    :: pdel(:,:)         ! layer thickness (Pa)
+
+        real(kind_phys), intent(inout) :: T(:,:)            ! air temperature (K)
+        real(kind_phys), intent(inout) :: qv(:,:)           ! specific humidity Q (kg/kg)
+
+        real(kind_phys), intent(out)   :: relhum(:,:)       ! relative humidity
+        real(kind_phys), intent(out)   :: precl(:)             ! large-scale precipitation rate (m/s)
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.F90]: Cleaned up!

File: tj2016/tj2016_precip.meta [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  dimensions = ()
+  intent = in
+[ rh2o ]
+  standard_name = gas_constant_of_water_vapor
+  units = J kg-1 K-1
+  type = real | kind = kind_phys
+  dimensions = ()
+  intent = in
+[ epsilo ]
+  standard_name = ratio_of_h2o_to_dry_air_molecular_weights
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.meta]: Updated!

File: tj2016/tj2016_precip.meta [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+[ qv ]
+  standard_name = water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water
+  units = kg kg-1
+  advected = True
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent, vertical_layer_dimension)
+  intent = inout
+[ relhum ]
+  standard_name = relative_humidity
+  units = 1
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.meta]: Updated!

File: tj2016/tj2016_precip.meta [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent, vertical_layer_dimension)
+  intent = out
+[ precl ]
+  standard_name = stratiform_precipitation_rate
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.meta]: Updated!

File: tj2016/tj2016_sfc_pbl_hs.F90 [hunk 12]
```diff
@@ -0,0 +1,480 @@
+module TJ2016_sfc_pbl_hs
+    use ccpp_kinds, only: kind_phys
+    use shr_const_mod, only: pi => shr_const_pi
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.F90]: Removed and updated API And meta.

File: tj2016/tj2016_sfc_pbl_hs.F90 [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+  
+    implicit none
+    private
+    save
+
+    public :: tj2016_sfc_pbl_hs_run
+
+CONTAINS
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.F90]: Lowercased!

File: tj2016/tj2016_sfc_pbl_hs.F90 [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+    !------------------------------------------------
+
+    integer,  intent(in)           :: ncol                ! number of columns
+    integer,  intent(in)           :: pver                ! number of vertical levels
+    integer,  intent(in)           :: pverp               ! number of vertical levels + 1
+    integer,  intent(in)           :: index_surface       ! index of surface (layer)
+    integer,  intent(in)           :: index_surface_interface ! index of surface (interface)
+
+    real(kind_phys), intent(in)    :: gravit              ! g: gravitational acceleration (m/s2)
+    real(kind_phys), intent(in)    :: cappa               ! Rd/cp
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.F90]: Updated!

File: tj2016/tj2016_sfc_pbl_hs.F90 [hunk 15]
```diff
@@ -0,0 +1,480 @@
...
+    real(kind_phys), intent(in)    :: rairv(:,:)          ! Rd: dry air gas constant (J/K/kg)
+    real(kind_phys), intent(in)    :: cpairv(:,:)         ! cp: specific heat of dry air (J/K/kg)
+    real(kind_phys), intent(in)    :: latvap              ! L: latent heat of vaporization (J/kg)
+    real(kind_phys), intent(in)    :: rh2o                ! Rv: water vapor gas constant (J/K/kg)
+    real(kind_phys), intent(in)    :: epsilo              ! Rd/Rv: ratio of h2o to dry air molecular weights
+    real(kind_phys), intent(in)    :: rhoh2o              ! density of liquid water (kg/m3)
+    real(kind_phys), intent(in)    :: zvirv(:,:)          ! (rh2o/rair) - 1, needed for virtual temperaturr
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.F90]: Corrected!

File: tj2016/tj2016_sfc_pbl_hs.meta [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+  dimensions = ()
+  intent = in
+[ rh2o ]
+  standard_name = gas_constant_of_water_vapor
+  units = J kg-1 K-1
+  type = real | kind = kind_phys
+  dimensions = ()
+  intent = in
+[ epsilo ]
+  standard_name = ratio_of_h2o_to_dry_air_molecular_weights
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.meta]: Fixed!

File: tj2016/tj2016_sfc_pbl_hs.meta [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+  units = Pa
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent)
+  intent = out
+[ tauy ]
+  standard_name = surface_y_momentum_flux_for_coupling
+  units = Pa
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent)
+  intent = out
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.meta]: Changed!

File: tj2016/tj2016_precip.meta [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+  type = character | kind = len=512
+  dimensions = ()
+  intent = out
+[ errflg ]
+  standard_name = ccpp_error_code
+  long_name = Error flag for error handling in CCPP
+  units = 1
+  type = integer
+  dimensions = ()
+  intent = out
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_precip.meta:142]: Added!

File: tj2016/tj2016_sfc_pbl_hs.meta [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+  intent = inout
+  advected = True
+[ shflx ]
+  standard_name = surface_upward_sensible_heat_flux_for_coupling
+  units = W m-2
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent)
+  intent = out
+[ lhflx ]
+  standard_name = surface_upward_latent_heat_flux
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply tj2016/tj2016_sfc_pbl_hs.meta:196]: @mwaxmonsky Looking at the updated spreadsheet it appears that the standard name for `lhflx` is correct as-is, but the standard name for `shflx` should be just `surface_upward_sensible_heat_flux`.  So if we could drop the `for_coupling` from the `shflx` standard name that would be ideal.  Thanks!

[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tj2016/tj2016_sfc_pbl_hs.meta:196]: Updated!

**Review by peverwhee — APPROVED** [review]
looks good to me!

**Pull request merged**